### DISCUSSION TITLE
🎲 Phase 23.1 Show & Tell — UncertaintyQuantifier: Bayesian uncertainty estimation & epistemic-aleatoric decomposition

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#535] [Show and tell]
# 🎲 Phase 23.1 — UncertaintyQuantifier: Bayesian Uncertainty Estimation & Epistemic-Aleatoric Decomposition

## Why Uncertainty Quantification Matters for AGI

Every prediction an AGI system makes carries **hidden uncertainty**. A planning module that treats a 60% confidence prediction the same as a 99% one will make catastrophically different decisions. Uncertainty quantification (UQ) is the **epistemic immune system** of autonomous intelligence — it tells the system *what it doesn't know* and *why*.

The `UncertaintyQuantifier` provides:
- **Calibrated confidence** — raw model outputs are notoriously overconfident; UQ corrects this
- **Epistemic-aleatoric decomposition** — separating *reducible* from *irreducible* uncertainty enables targeted action
- **Decision-theoretic grounding** — downstream modules (RiskAssessor, DecisionOrchestrator) need quantified uncertainty, not point estimates

## Architecture & Data Flow

```
┌─────────────────────────────────────────────────────────────────────┐
│                    UncertaintyQuantifier Pipeline                    │
├─────────────────────────────────────────────────────────────────────┤
│                                                                     │
│  Predictions ──► EnsembleDisagreement ──► decompose()               │
│  (N models)      (variance across          │                        │
│                   ensemble members)         ▼                       │
│                                    UncertaintyEstimate              │
│                                    ├─ epistemic: float  (model)     │
│                                    ├─ aleatoric: float  (data)      │
│                                    ├─ total: float                  │
│                                    └─ confidence: float             │
│                                             │                       │
│                                             ▼                       │
│                                      calibrate()                    │
│                                    (Platt / Isotonic /              │
│                                     Temperature / Beta)             │
│                                             │                       │
│                                             ▼                       │
│                                   CalibratedEstimate                │
│                                    ├─ calibrated_conf: float        │
│                                    ├─ ece_score: float              │
│                                    └─ reliability_curve: List       │
│                                             │                       │
│                              ┌──────────────┼──────────────┐        │
│                              ▼              ▼              ▼        │
│                        RiskAssessor   ScenarioPlanner  Decision     │
│                          (23.2)         (23.3)       Orchestrator   │
│                                                        (23.5)       │
└─────────────────────────────────────────────────────────────────────┘
```

## Uncertainty Type Decomposition

| Type | Source | Reducible? | Action | Example |
|------|--------|-----------|--------|---------|
| **Epistemic** | Model uncertainty — limited training data, model misspecification | ✅ Yes — more data, better model | Trigger active learning, data gathering, ask for clarification | "I haven't seen inputs like this before" |
| **Aleatoric** | Inherent noise in the data-generating process | ❌ No — irreducible | Trigger robust/hedged strategies, widen confidence intervals | "This outcome is inherently stochastic" |
| **Total** | `sqrt(epistemic² + aleatoric²)` | Partial | Overall uncertainty budget for decision thresholds | Combined risk signal |

**Decomposition method**: Ensemble disagreement naturally separates the two:
- **Epistemic** ≈ variance of ensemble *means* (models disagree → they lack knowledge)
- **Aleatoric** ≈ mean of ensemble *variances* (models agree on spread → data is noisy)

## Calibration Methods Comparison

| Method | Mechanism | Parameters | Strengths | Weaknesses |
|--------|-----------|-----------|-----------|------------|
| **Platt Scaling** | Logistic regression on logits → calibrated probabilities | 2 (slope + intercept) | Simple, effective for binary; low overfit risk | Assumes sigmoid shape; poor for multi-class |
| **Isotonic Regression** | Non-parametric monotonic fit on holdout set | O(n) bins | No shape assumption; adapts to any calibration curve | Can overfit with small holdout; requires sorting |
| **Temperature Scaling** | Single scalar T divides logits before softmax | 1 (temperature) | Minimal params; preserves ranking; multi-class native | Assumes uniform miscalibration across classes |
| **Beta Calibration** | Beta distribution CDF fit on [0,1] probabilities | 3 (a, b, c) | Flexible shape; handles both under/over-confidence | More params; needs careful regularisation |

The `UncertaintyQuantifier` supports all four via `CalibrationMethod` enum, with `Temperature Scaling` as default for multi-class and `Platt Scaling` for binary tasks.

## Integration with Phase 23 Pipeline

```
Phase 22 CreativeOrchestrator
        │
        ▼
Phase 23: Uncertainty-Aware Decision Making
├── 23.1 UncertaintyQuantifier  ◄── THIS COMPONENT
│     └── UncertaintyEstimate + CalibratedEstimate
│              │
│              ├──► 23.2 RiskAssessor
│              │     └── risk_score = f(uncertainty, consequence)
│              │
│              ├──► 23.3 ScenarioPlanner
│              │     └── scenario weights ∝ calibrated confidence
│              │
│              ├──► 23.4 ContingencyGenerator
│              │     └── fallback plans where epistemic > threshold
│              │
│              └──► 23.5 DecisionOrchestrator
│                    └── expected utility maximisation under uncertainty
```

## Prometheus Metrics

| Metric | Type | Description |
|--------|------|-------------|
| `uq_estimates_total` | Counter | Total uncertainty estimates produced |
| `uq_epistemic_ratio` | Histogram | Epistemic / total uncertainty ratio (0–1) |
| `uq_calibration_ece` | Gauge | Expected Calibration Error after latest recalibration |
| `uq_recalibration_duration_seconds` | Histogram | Time to refit calibration model |
| `uq_ensemble_size` | Gauge | Current number of ensemble members |

**PromQL examples**:
```promql
# Alert: calibration drift — ECE exceeds 0.05
uq_calibration_ece > 0.05

# Ratio of predictions with high epistemic uncertainty (>0.7)
sum(rate(uq_epistemic_ratio_bucket{le="1.0"}[5m]) - rate(uq_epistemic_ratio_bucket{le="0.7"}[5m]))
  / sum(rate(uq_epistemic_ratio_count[5m]))

# P99 recalibration latency
histogram_quantile(0.99, rate(uq_recalibration_duration_seconds_bucket[15m]))
```

## Open Questions

1. **Ensemble size vs accuracy trade-off** — Literature suggests 5–10 members capture most epistemic variance (Lakshminarayanan et al., 2017), but Phase 23's computational budget is shared with ScenarioPlanner. Should we support adaptive ensemble sizing based on available compute?

2. **Online calibration drift** — If the prediction-outcome distribution shifts gradually, the calibration model can become stale. Should `recalibrate()` use an exponentially-weighted sliding window, or periodic batch re-fit? What's the right window size?

3. **Heteroscedastic aleatoric estimation** — Current design assumes homoscedastic aleatoric noise across the input space. For some domains (e.g., financial predictions), aleatoric uncertainty varies dramatically with input features. Should we add input-conditioned aleatoric heads to ensemble members?

---

*Phase 23.1 of the ASI-Build cognitive architecture. See issue #530 for the full specification.*
